MRG: Revamp HTML reprs of `Raw`, `Epochs`, `Evoked`, and `Info`

hoechenberger opened the pull request:
For example, the `Evoked` repr contained the number of channels, but failed to subtract or report any bads.

Also I believe reporting the number of sampling points is quite pointless , so I removed that as well.

The table layout wasn't up-to-date either.

`main` (striped layout only added by VS Code, actually not present in the tables we produce!)

<img width="605" alt="Screenshot 2024-04-28 at 00 36 50" src="https://github.com/mne-tools/mne-python/assets/2046265/4f9b9fc1-9aac-4f49-ac9e-e38c57de5ac2">

This PR:

<img width="605" alt="Screenshot 2024-04-28 at 00 41 39" src="https://github.com/mne-tools/mne-python/assets/2046265/eff59128-9611-47a4-bd87-ee704cb3c42d">


For comparison, `Epochs`:

<img width="323" alt="Screenshot 2024-04-28 at 00 36 15" src="https://github.com/mne-tools/mne-python/assets/2046265/1589b4d6-7024-4286-8dbe-1d8d07ecced6">


MWE:
```python
# %%
import mne

sample_dir = mne.datasets.sample.data_path()
sample_fname = sample_dir / "MEG" / "sample" / "sample_audvis_raw.fif"

raw = mne.io.read_raw_fif(sample_fname)
raw.crop(tmax=60)

events = mne.find_events(raw, stim_channel="STI 014")
raw.pick("eeg")

epochs = mne.Epochs(raw, events=events, preload=True)
epochs

# %%
evoked = epochs.average()
evoked
```

This one also
closes #12553

hoechenberger commented:
This PR should not be merged before #12550 goes in, after which we may need to update this one slightly to ensure consistency between Epochs and Evoked repr

agramfort commented:
you're aligning to the minimum info by dropping number of channels and time points. I find it useful to know the duration and how many channels are available.

agramfort commented:
yes I would have done that if you push for consistency.

> Message ID: ***@***.***>
>

hoechenberger commented:
Ok i'll take a look!

hoechenberger commented:
Ok, so I've now touched the `Raw`, `Epochs`, and `Evoked` HTML reprs to make things more consistent. Here's a comparison of what we have in `main` and what's in this PR branch now:

![Comparison](https://github.com/mne-tools/mne-python/assets/2046265/2e681826-e32d-47ec-a666-9cff1ecb6daf)

drammock commented:
I'm generally quite happy with the improved consistency. Thanks! Nitpicks:

- I find "digitized head and sensor positions" doesn't quite make sense (the headshape points aren't "head positions" in the sense that we normally use that term). I'd revert to "digitized points" 
- the "data kind" field is only in evoked; is that meant to communicate "avg" vs "stdev"?  Does it make sense to collapse that into one field with `nave` (something like "average of NN epochs" or "standard deviation of NN epochs")?

hoechenberger commented:
> I'm generally quite happy with the improved consistency. Thanks! Nitpicks:
> 
> * I find "digitized head and sensor positions" doesn't quite make sense (the headshape points aren't "head positions" in the sense that we normally use that term). I'd revert to "digitized points"

I was worried this could be confused with "sampling points" on the time axis … I had first thought of labeling it "digitized had shape and sensor positions", but it's too long. I'm okay with changing it back to "digitized points", but it's not my favorite – maybe if we brainstorm a little, we can come up with a better idea?

> * the "data kind" field is only in evoked; is that meant to communicate "avg" vs "stdev"? 

Yes, it was there before and I'm actually not sure if it's really needed? 

> Does it make sense to collapse that into one field with `nave` (something like "average of NN epochs" or "standard deviation of NN epochs")?

Good idea, I can do that!

hoechenberger commented:
@drammock Another thing we may need to address – these tables are quite long now, and I haven't added collapsible sections like we have for `Info`. Is that a problem for the tutorial rendering?

mscheltienne commented:
+1 for `digitized points`, unless someone comes up with a clearer alternative. Thanks for the upgrade, it looks great!

hoechenberger commented:
@mscheltienne I named the row now "Head shape and sensor digitization", and the value is then "xxx points"
WDYT?

drammock commented:
> @drammock Another thing we may need to address – these tables are quite long now, and I haven't added collapsible sections like we have for `Raw` (which essentially just prints out `Info`). Is that a problem for the tutorial rendering?

Ideally they'd have collapsible sections, yeah. Can you add them in this PR? If not please create a follow-up issue so someone can do it later.

hoechenberger commented:
> Ideally they'd have collapsible sections, yeah. Can you add them in this PR?

This then also raises the question of how much Info (pun intended) to include here

For Raw, the HTML repr currently is literally `Raw.info`.

For Epochs and Evoked, it's a mixture of some instance attributes plus some select parts I extract from `inst.info`.

If we add collapsibles, we could add the entire Info, but I'm sure @cbrnr would hate it :)

Re collapsibles, I like the idea _per se_, but the current implementation for Info is visually unpleasant, since each section gets its own table and hence the columns don't align. I wonder if we can find a better solution.

hoechenberger commented:
> Re collapsibles, I like the idea _per se_, but the current implementation for Info is visually unpleasant, since each section gets its own table and hence the columns don't align. I wonder if we can find a better solution.

I will look into this and propose a solution

cbrnr commented:
I don't like the idea to misuse standard reprs of Python objects. You are only modifying HTML reprs, which I think is fine (and I don't use them anyway, so I really have no opinion on what should and should not be part of these tables).

hoechenberger commented:
Got you! @cbrnr

larsoner commented:
Looks like there are plenty of folks chiming in here with ideas so as long as you all converge I don't think I need to look, just wanted to say thanks for looking into this!

hoechenberger commented:
@larsoner If you have any thoughts on potentially a better label for "Digitized points", I'd love to hear it!

Maybe it's just me but in my ears it sounds kind of ambiguous / unspecific or even confusing when this information is presented in a summary table where we also display the number of time-series data sampling points

Any input welcome

hoechenberger commented:
> > Re collapsibles, I like the idea _per se_, but the current implementation for Info is visually unpleasant, since each section gets its own table and hence the columns don't align. I wonder if we can find a better solution.
> 
> I will look into this and propose a solution

I've found a solution, but it may take a day or two to implement. Will report back once I've had the time to do that

hoechenberger commented on doc/changes/devel/12583.newfeature.rst:
```diff
@@ -0,0 +1,4 @@
+The HTML representations of :class:`~mne.Raw`, :class:`~mne.Epochs`,
```
```suggestion
The HTML representations of :class:`~mne.io.Raw`, :class:`~mne.Epochs`,
```

hoechenberger commented on mne/_fiff/meas_info.py:
```diff
@@ -1900,7 +1900,7 @@ def _repr_html_(
         meas_date = self.get("meas_date")
         if meas_date is not None:
-            meas_date = meas_date.strftime("%B %d, %Y  %H:%M:%S") + " GMT"
+            meas_date = meas_date.strftime("%B %d, %Y  %H:%M:%S") + " UTC"
```
cc @cbrnr

hoechenberger commented on mne/_fiff/meas_info.py:
```diff
@@ -10,10 +10,12 @@
 import datetime
 import operator
 import string
+import uuid
```
we create unique HTML class names each time we produce a new HTML repr

hoechenberger commented on doc/changes/devel/12583.apichange.rst:
```diff
@@ -0,0 +1,2 @@
+:attr:`mne.Info.ch_names` will now return an empty list instead of raising a `KeyError` if no channels
```
@larsoner @drammock This reference apparently doesn't work -- do you know why? ch_names is a property of Info

drammock replied:
AFAIK it should work: https://www.sphinx-doc.org/en/master/usage/domains/python.html#role-py-attr

hoechenberger replied:
Maybe it's something about the namespace ...

The doc build fails with:

> /home/circleci/project/doc/<rst_prolog>:30: WARNING: py:attr reference target not found: mne.Info.ch_names [ref.attr]

hoechenberger commented:
This is ready for review

I think a good way to explore these changes is by browsing through the Report tutorial, because the different examples there demonstrate the HTML repr for Raw, Epochs, and Evoked:

https://output.circle-artifacts.com/output/job/ce2dba84-e894-4f81-b168-a0b0af9807be/artifacts/0/html/auto_tutorials/intro/70_report.html#adding-raw-data

Note that for the documentation build, the repr sections are collapsed; during interactive use or when users create reports, all sections are expanded.

I consolidated repr sections across Info, Raw, Epochs, and Evoked.

Many helper functions were converted to Jinja2 filters, which allows us to do less "preparatory work" inside the `_repr_html_()` methods (e.g. for generating human-readable strings of datetimes); instead, this can now all be taken care of inside the Jinja2 templates, with only a few exceptions.

This PR contains one breaking API change: `Info.ch_names` used to raise an exception if `Info["ch_names"]` was empty; now, it returns an empty list. This made certain operations for me easier. Since this should be a very rare edge case, I don't believe a deprecation cycle is necessary.

cc @agramfort

agramfort commented:
honestly having to click on each section to see something is to me an
overkill. It forces me to use
my mouse and while before shift+enter is enough to see what I want. Could
we find a middle ground here?
eg you see the important info by default and you can click to have the full
details? or expand certain sections
by default?

nit: the + sign is not to me the most common. It's more classic to have >
and that become \/ when it's expanded.

> Message ID: ***@***.***>
>

hoechenberger commented:
@agramfort There might be a misunderstanding here, the clicking is only required in the documentation as is already the case with what we have in `main`

If instead you're using your Python Interactive window or creating a Report, everything is expanded by default, so no clicking is necessary!

Try this:

```python
# %%
import mne

sample_dir = mne.datasets.sample.data_path()
sample_fname = sample_dir / "MEG" / "sample" / "sample_audvis_raw.fif"

raw = mne.io.read_raw_fif(sample_fname)
raw.crop(tmax=60)

events = mne.find_events(raw, stim_channel="STI 014")
raw.pick("eeg")

# %%
epochs = mne.Epochs(raw, events=events, preload=True)
epochs

# %%
evoked = epochs.average()
evoked

# %%
raw.info
```

hoechenberger commented:
> the + sign is not to me the most common. It's more classic to have > and that become \/ when it's expanded.

Yep, I will try to find a replacement

hoechenberger commented on mne/html_templates/repr/static/repr.css:
```diff
@@ -0,0 +1,64 @@
+  background-color: transparent;
+  border-color: transparent;
+}
+
+.collapse-uncollapse-caret.hidden {
+  display: none;
+}
```
```suggestion
```

hoechenberger commented:
@agramfort 
> > the + sign is not to me the most common. It's more classic to have > and that become / when it's expanded.
> 
> Yep, I will try to find a replacement

This has been resolved, see:
https://output.circle-artifacts.com/output/job/9a976a2f-b617-4f47-9f43-8c4c1d998eae/artifacts/0/html/auto_tutorials/intro/70_report.html#adding-raw-data

hoechenberger commented:
Latest rendered Report tutorial:
https://output.circle-artifacts.com/output/job/6a2f0a7d-04bb-4581-8c01-0b2ef7bdbbbe/artifacts/0/html/auto_tutorials/intro/70_report.html#adding-raw-data

CI errors seem unrelated

hoechenberger commented:
Dang, the carets (collapse / uncollapse buttons) don't show up on my iPad. Need to investigate.

hoechenberger commented:
I think this is ready for review now. I addressed @agramfort's concerns regarding the collapse / uncollapse buttons (they are carets now instead of plus / minus symbols). It seems to look ok (not perfect) in Reports on both desktop and my iPad, and in the VS Code Python Interactive console.

The segments of the repr are all expanded by default, and only collapsed in the built documentation (as is the case with our stable docs).

Rendered Report tutorial:
https://output.circle-artifacts.com/output/job/bb7905a0-af6b-4d52-a43e-b196baa58e63/artifacts/0/html/auto_tutorials/intro/70_report.html#adding-raw-data

mscheltienne reviewed (approved):
Looks great, the length of `Head shape and sensor digitization` is not bothering me. 
Tested the following on computer:

```python
from mne import make_fixed_length_epochs
from mne.datasets import sample
from mne.io import read_raw_fif

folder = sample.data_path() / "MEG" / "sample"
raw = read_raw_fif(folder / "sample_audvis_raw.fif", preload=True)
raw

epochs = make_fixed_length_epochs(raw)
epochs

evoked = epochs.average()
evoked
```

Thanks @hoechenberger

hoechenberger commented:
Thanks for your feedback, @mscheltienne!

> Looks great, the length of `Head shape and sensor digitization` is not bothering me.

I misread this as "**is** bothering me" and came up with a shorter proposal:
"Head & Sensor Digitization" :) Thoughts?

mscheltienne commented:
It's actually a bit better :laughing:

drammock commented on mne/epochs.py:
```diff
@@ -2085,7 +2080,15 @@ def _repr_html_(self):
+            inst=self,
+            filenames=(
+                [Path(self.filename).name]
+                if hasattr(self, "filename") and self.filename is not None
```
```suggestion
                if getattr(self, "filename", None) is not None
```

hoechenberger replied:
Good idea

drammock commented on mne/evoked.py:
```diff
@@ -467,14 +468,15 @@ def __repr__(self):  # noqa: D105
+            inst=self,
+            filenames=(
+                [Path(self.filename).name]
+                if hasattr(self, "filename") and self.filename is not None
```
```suggestion
                if getattr(self, "filename", None) is not None
```

drammock commented on mne/html_templates/_templates.py:
```diff
@@ -1,10 +1,105 @@
+
+def _dt_to_str(dt: datetime.datetime) -> str:
+    """Convert a datetime object to a human-reaable string representation."""
+    return dt.strftime("%B %d, %Y    %H:%M:%S") + " UTC"
```
does `%Z` not work here?
https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes

hoechenberger replied:
This could work, yes
I can try it out

hoechenberger replied:
This seems to work!

While we're at it: WDYT about a date representation like this:

```
2002-12-03 at 19:01:10 UTC
```
(YYYY-MM-DD)

drammock replied:
> YYYY-MM-DD

That is my personal preferred date format, so I'm fine with it :)
It's also what ISO 8601 stipulates, so it's as good a choice as any

hoechenberger replied:
@drammock I addressed this in [fd976a5](https://github.com/mne-tools/mne-python/pull/12583/commits/fd976a53c5ce2386f70dcc1664a52e3cb256c9ad)

I didn't implement "full" ISO-8601, because it's not super human-readable when you have a datetime string; but I took inspiration from that standard. Please LMK what you think

drammock reviewed (approved):
it looks beautiful @hoechenberger 😍 And the refactoring of the Jinja templates resolves a big pain point --- it will be much easier to make other object types have consistent HTML reprs now too.

I left a couple minor inline comments/suggestions, here are three other tiny suggestions/ideas from looking at the rendered output:

1. In the "general" section, "Data type" feels like not quite the right name (to me "data type" means float, int, complex, etc). Maybe "object type"?
2. I'm very slightly bothered by the title of the "Data" section (seems too generic of a title). At first I thought "Times" (to parallel the "channels" section) would work instead --- for raw and epochs, the "data" section is all about time-domain things (duration, sfreq, events; events is not strictly about time, but if you squint it could make sense under a "times" heading).  Sadly this doesn't work for evokeds, because there we also get "aggregation" and "condition" entries. I think it's OK to leave as "data" for now unless someone has a quick idea about a better alternative.
3. I think it's confusing for "good channels" to say "59 EEG" (a count) and "bad channels" to say "EEG 053" (a channel name that looks *a lot* like a count of channels of a particular type).  A couple possible repairs:
    a. channel names are always enquoted
    b. when there are few enough channels to be worth writing them individually, write a count followed by names in parentheses (e.g., `bad channels: 1 (EEG 053)`).

All of these are small enough that they can be done later / are not blockers for me

hoechenberger commented on mne/html_templates/_templates.py:
```diff
@@ -1,10 +1,105 @@
+
+
+def _dt_to_str(dt: datetime.datetime) -> str:
+    """Convert a datetime object to a human-reaable string representation."""
```
```suggestion
    """Convert a datetime object to a human-readable string representation."""
```

agramfort reviewed (commented):
thanks @hoechenberger !

hoechenberger commented:
@drammock FYI when I started with this work, I named the `Data` section `Recording`, which seemed nice until I realized for an `Info` object, it seemed kinda odd, because it doesn't have any "recorded" data attached.

**That said,** one could argue that a Data section is just as inappropriate there. 

So what do you say, shall we rename Data -> Recording? I can push a commit so you can take a look at the rendered reprs to get a better idea of what it "feels" like

hoechenberger commented:
> * In the "general" section, "Data type" feels like not quite the right name (to me "data type" means float, int, complex, etc). Maybe "object type"?

I have addressed this in aacb2c4 and renamed it to "MNE object time", WDYT?

hoechenberger commented:
> 3\. I think it's confusing for "good channels" to say "59 EEG" (a count) and "bad channels" to say "EEG 053" (a channel name that looks _a lot_ like a count of channels of a particular type).  A couple possible repairs:
>     a. channel names are always enquoted
>     b. when there are few enough channels to be worth writing them individually, write a count followed by names in parentheses (e.g., `bad channels: 1 (EEG 053)`).

I'll look into this. Thanks to the refactoring I've done here, I believe now I only have to touch the `_format_channels()` function (i.e., all is in one place). Neat!

hoechenberger commented:
> @drammock FYI when I started with this work, I named the `Data` section `Recording`, which seemed nice until I realized for an `Info` object, it seemed kinda odd, because it doesn't have any "recorded" data attached.
> 
> **That said,** one could argue that a Data section is just as inappropriate there.
> 
> So what do you say, shall we rename Data -> Recording? I can push a commit so you can take a look at the rendered reprs to get a better idea of what it "feels" like

I renamed Data to Acquisition

> @drammock FYI when I started with this work, I named the `Data` section `Recording`, which seemed nice until I realized for an `Info` object, it seemed kinda odd, because it doesn't have any "recorded" data attached.
> 
> **That said,** one could argue that a Data section is just as inappropriate there.
> 
> So what do you say, shall we rename Data -> Recording? I can push a commit so you can take a look at the rendered reprs to get a better idea of what it "feels" like

I renamed Data to Acquisition

drammock commented:
> I renamed Data to Acquisition

that's an improvement I think. The number averaged in an Evoked isn't really "acquisition" but I think it's OK.

hoechenberger commented:
I think this is ready to be merged. It was quite challenging to build something that works and looks okay-ish in VS Code, Jupyter, and in Report. For example, I wanted to use the [Popover API](https://developer.mozilla.org/en-US/docs/Web/API/Popover_API) to create popups, but VS Code would render the Python Console input widgets above my popover content, while Jupyter wouldn't. Problems such as these.

Anyway, here are some examples:

### VS Code Python Interactive, light theme
https://github.com/mne-tools/mne-python/assets/2046265/5e788fc2-e52f-4645-b28f-2f234659bd93

### VS Code Python Interactive, dark theme
<img width="612" alt="Screenshot 2024-05-19 at 09 15 33" src="https://github.com/mne-tools/mne-python/assets/2046265/95f3d2c9-6f09-4fef-a7bc-38a3f582da53">

### Report, light theme
https://github.com/mne-tools/mne-python/assets/2046265/c75d4a86-628f-4b06-8d0e-bcb97c7d5c89

### Report, dark theme
<img width="999" alt="Screenshot 2024-05-19 at 09 58 33" src="https://github.com/mne-tools/mne-python/assets/2046265/73414726-fea5-4f17-b35d-c7941f229a6c">

### Jupyter Lab, light theme
<img width="475" alt="Screenshot 2024-05-19 at 10 00 49" src="https://github.com/mne-tools/mne-python/assets/2046265/03cb8f50-442c-4792-8883-e8569ac717fc">

### Jupyter Lab, dark theme
<img width="475" alt="Screenshot 2024-05-19 at 10 01 09" src="https://github.com/mne-tools/mne-python/assets/2046265/37f29ff3-f93f-43c1-b9da-04b5e8e44147">

As you can see, there's a problem with dark theme support in Report and Jupyter Lab (but not in VS Code Python Interactive), as the collapse/uncollapse caret is simply changed to a white color, which of course only works if the background is dark.

I think for now this is good enough and I will try to improve dark theme support in some follow-up PRs; I've been wanting to add a dark theme to Report for a while now anyway.

Also the examples demonstrate that now, you can click on the entire section header row (not just on the carets) to collapse/uncollapse a section.

Rendered tutorial: https://output.circle-artifacts.com/output/job/d78b2fa9-df94-4d24-ad32-1b2806c11212/artifacts/0/html/auto_tutorials/intro/70_report.html#adding-raw-data

hoechenberger commented:
Oh no, I'm realizing just now that the section headers in Jupyter Lab are right-justified, while they should be left-justified. I need to look into this.

… you see, it's stuff like that that's eating away my time and nerves :D

hoechenberger commented:
I fixed the text alignment of the section headers in Jupyter Lab (was right-justified in the screenshots above; should be left-justified like below):

<img width="475" alt="Screenshot 2024-05-19 at 10 25 31" src="https://github.com/mne-tools/mne-python/assets/2046265/83d0b4e1-0253-438f-9c56-e162c59a9424">

hoechenberger commented:
I realized there's still some issues with the text alignment in Jupyter Lab, need to fix those first.

hoechenberger commented:
I think it's all fixes now in Jupyter Lab!

<img width="475" alt="Screenshot 2024-05-19 at 10 35 43" src="https://github.com/mne-tools/mne-python/assets/2046265/29803bab-bd0e-4601-80a6-c8ebcde50df5">

hoechenberger commented:
Ha! In my above "Jupyter Lab, dark theme" tests I only set my OS theme to dark, but not Jupyter. However, in CSS I'm of course querying the system. Setting Jupyter Lab to follow the system theme, I get the following in dark mode:

<img width="475" alt="Screenshot 2024-05-19 at 21 03 00" src="https://github.com/mne-tools/mne-python/assets/2046265/9fd1b790-658c-4a78-b4a8-d6d9c27ce248">

which is just what we want.

So the only thing that's impaired for now in dark mode is Report.

larsoner commented:
@mscheltienne feel free to review/merge if you're happy, or if you don't have time I can look

mscheltienne commented:
I'm away until Thursday, I can look then.

hoechenberger commented:
@larsoner If you want, go ahead and merge; and I can address comments from @mscheltienne in a follow-up PR (I expect we'll have to do a few of those once we start using the new code)

drammock commented:
let's get this in, I agree that follow-up PRs are fine

mscheltienne commented:
Looks good to me, the only comment I have is on the pop-up you get when clicking on the number of channels (good or bad); on VSCode I don't find the notification very easy to use. But that's a very minor comment for a great refactor!
